03-task-manager/starter/quizAnswers.txt

```diff
@@ -0,0 +1,12 @@
+1.Quickly explain how middleware is used in an Express application.
+    Middleware in Express is a function that handles the request-response cycle and performs tasks such as validation, authentication, data parsing, etc.
+    It can be added to an application using app.use() or app.route() and executed in a sequence.
```

Comment:
`app.use()` is how we add middleware in Express. 

`app.route()` is more specifically a way to create ["routes".](https://expressjs.com/en/guide/routing.html#app-route) So that one is not _exactly_ a middleware, it's just a way to define a set of related handlers for requests under a common path (eg: in this week's assignment we define a set of routes under `/api/v1/tasks` and another set of routes under `/api/v1/tasks/:id`.  Usually, when we talk about middleware we mean specifically the things that happen in the "middle" of accepting a request but before sending the response.

You can also add middleware only for specific paths by passing in the path as the first argument to `app.use()` like this:

```
app.use('/user/:id', (req, res, next) => {
  console.log('Request Type:', req.method)
  next()
})
```

This will only log the request method if the browser navigates to the URL with a path like `/user/123`

Instead of writing the function right in the `app.use()` , you could also first define the function and then pass it in. And you can pass in as many middleware functions as you want:

```
function logOriginalUrl (req, res, next) {
  console.log('Request URL:', req.originalUrl)
  next()
}

function logMethod (req, res, next) {
  console.log('Request Type:', req.method)
  next()
}

app.use('/user/:id', logOriginalUrl, logMethod)
```

This will log the request URL and request type for all requests to URLs liek `/user/123` before letting the route handler actually build the response.

You can also define the middleware only on a specific set of routes like this:

```
const express = require('express');
const {
  getAllTasks,
  createTask,
  getTask,
  updateTask,
  deleteTask,
} = require('../controllers/tasks');

const router = express.Router();

router.use(logOriginalUrl, logMethod)
router.route('/').get(getAllTasks).post(createTask);
router.route('/:id').get(getTask).patch(updateTask).delete(deleteTask);
```

This will only do the `logOriginalUrl` and `logMethod` middlewares for any of these tasks routes but not for any other path in the Express application.

Lastly, you could even define middleware specifically on a single route/path handler:

```
const logStuff = [logOriginalUrl, logMethod]
router.route('/').get(logStuff, getAllTasks).post(createTask);
```

This will only do the logging on the `/api/v1/tasks` endpoint and not on any of the other ones.